include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1137,49 +1136,114 @@ struct __early_exit_find_or
+    // Starts main work...
 
-    // TODO: find a way to generalize getting of reliable work-group size
-    std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
-#if _ONEDPL_COMPILE_KERNEL
-    auto __kernel = __internal::__kernel_compiler<_FindOrKernel>::__compile(__exec);
-    __wgroup_size = ::std::min(__wgroup_size, oneapi::dpl::__internal::__kernel_work_group_size(__exec, __kernel));
-#endif
+    using __result_and_scratch_storage_t = __result_and_scratch_storage<_ExecutionPolicy, _AtomicType>;
+    __result_and_scratch_storage_t __result_storage(__exec, 0);
 
-#if _ONEDPL_FPGA_EMU
-    // Limit the maximum work-group size to minimize the cost of work-group reduction.
-    // Limiting this also helps to avoid huge work-group sizes on some devices (e.g., FPGU emulation).
-    __wgroup_size = std::min(__wgroup_size, (std::size_t)2048);
+    // Calculate the number of elements to be processed by each work-item.
+    const auto __iters_per_work_item = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __n_groups * __wgroup_size);
+
+    // main parallel_for
+    auto __event_id = __exec.queue().submit([&](sycl::handler& __cgh) {
+        oneapi::dpl::__ranges::__require_access(__cgh, __rngs...);
+        auto __result_acc = __result_storage.__get_result_acc(__cgh);
+
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+        __cgh.use_kernel_bundle(__kernel.get_kernel_bundle());
 #endif
-    auto __max_cu = oneapi::dpl::__internal::__max_compute_units(__exec);
+        __cgh.parallel_for<KernelName>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+            __kernel,
+#endif
+            sycl::nd_range</*dim=*/1>(sycl::range</*dim=*/1>(__n_groups * __wgroup_size),
+                                      sycl::range</*dim=*/1>(__wgroup_size)),
+            [=](sycl::nd_item</*dim=*/1> __item_id) {
+                auto __local_idx = __item_id.get_local_id(0);
+
+                // 1. Set initial value to local found state
+                _AtomicType __found_local = __init_value;
+
+                // 2. Find any element that satisfies pred
+                //  - after this call __found_local may still have initial value:
+                //    1) if no element satisfies pred;
+                //    2) early exit from sub-group occurred: in this case the state of __found_local will updated in the next group operation (3)
+                __pred(__item_id, __rng_n, __iters_per_work_item, __n_groups * __wgroup_size, __found_local,
+                       __brick_tag, __rngs...);
+
+                // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                // or update state with __dpl_sycl::__any_of_group (for the __parallel_or_tag)
+                // inside all our group items
```

Comment:
This comment looks outdated. I would remove it.

Reply:
Why?

Reply:
I misread the comment in line 1197. I would still suggest removing this line here and in the multi WG kernel since we explain that these are group operations already in line 1194 and 1196.

Reply:
Is it really significant?

Reply:
No but it would shorten the code. Your decision.

Reply:
Let's save as is.